File: haskell/src/SpaceTraders/Model/Route.hs
module SpaceTraders.Model.Route
  ( Route(..)
  , RouteEndpoint(..)
  ) where

import Data.Aeson
import Data.Time
import GHC.Generics
import qualified Data.Text as T

data Route = Route { arrival :: UTCTime
                   , departureTime :: UTCTime
                   , destination :: RouteEndpoint
                   , origin :: RouteEndpoint
                   } deriving (FromJSON, Generic, Show, ToJSON)

data RouteEndpoint = RouteEndpoint { routeEndpointType :: T.Text
                                   , symbol :: T.Text
                                   , systemSymbol :: T.Text
                                   , x :: Int
                                   , y :: Int
                                   } deriving (Generic, Show)
instance FromJSON RouteEndpoint where
  parseJSON = withObject "RouteEndpoint" $ \o ->
    RouteEndpoint <$> o .: "type"
                  <*> o .: "symbol"
                  <*> o .: "systemSymbol"
                  <*> o .: "x"
                  <*> o .: "y"
instance ToJSON RouteEndpoint where
  toJSON (RouteEndpoint t s ss xx yy) = object [ "type" .= t
                                               , "symbol" .= s
                                               , "systemSymbol" .= ss
                                               , "x" .= xx
                                               , "y" .= yy ]

File: golang/pkg/model/route.go
package model

import "time"

type Route struct {
	Arrival       time.Time     `json:"arrival"`
	DepartureTime time.Time     `json:"departureTime"`
	Destination   RouteEndpoint `json:"destination"`
}

type RouteEndpoint struct {
	Type         string `json:"type"`
	Symbol       string `json:"symbol"`
	SystemSymbol string `json:"systemSymbol"`
	X            int    `json:"x"`
	Y            int    `json:"y"`
}
